### PR TITLE
Fix jvm uswgi metrics

### PR DESCRIPTION
**Pull request by gnufied**
This implements reporting stats of Java applications which report their stats in a format supported by Dropwizard.

### CODE REVIEW

File: src/fullerite/collector/sample.json [hunk 1]
```diff
@@ -0,0 +1,997 @@
+{
```
Comment (zachm) [inline src/fullerite/collector/sample.json]: This file doesn't belong with the code... I'd be open to other places to keep it, where do you think?

File: src/fullerite/collector/nerve_uwsgi.go [hunk 2]
```diff
@@ -241,26 +244,43 @@ func convertToMetrics(metricMap *map[string]map[string]interface{}, metricType s
...
+	// only add things that have a numeric base
+	switch value.(type) {
+	case float64:
+		m.Value = value.(float64)
+	case int:
+		m.Value = float64(value.(int))
+	default:
+		return m, false
+	}
+	return m, true
```
Comment (zachm) [inline src/fullerite/collector/nerve_uwsgi.go]: this should do nothing since you included a default case

File: src/fullerite/collector/dropwizard_collector.go [hunk 3]
```diff
@@ -0,0 +1,162 @@
+package collector
```
Comment (zachm) [inline src/fullerite/collector/dropwizard_collector.go]: I'd like this renamed. It's not a Fullerite collector in and of itself (afaict) but is rather a set of functions used by NerveUWSGI.

There's also some spaghetti-type function calls where the two call into one another. It's better practice to extract those into a common library file if at all possible.

File: src/fullerite/collector/nerve_uwsgi.go [hunk 4]
```diff
@@ -241,26 +244,43 @@ func convertToMetrics(metricMap *map[string]map[string]interface{}, metricType s
...
-				results = append(results, m)
-			}
+		tempResults := metricFromMap(&metricData, metricName, metricType)
+		results = append(results, tempResults...)
+	}
+
+	return results
+}
+
+func metricFromMap(metricMap *map[string]interface{}, metricName string, metricType string) []metric.Metric {
```
Comment (zachm) [inline src/fullerite/collector/nerve_uwsgi.go]: Please add comments to this function, and all the other new ones.

File: src/fullerite/collector/dropwizard_collector.go [hunk 5]
```diff
@@ -0,0 +1,162 @@
...
+
+func collectGauge(jsonMap map[string]interface{}, metricName []string, metricType string) []metric.Metric {
+	compositeMetricName := strings.Join(metricName, ".")
+	return metricFromMap(&jsonMap, compositeMetricName, metricType)
+}
+
+func collectHistogram(jsonMap map[string]interface{}, metricName []string, metricType string) []metric.Metric {
+	if _, ok := jsonMap["count"]; ok {
+		compositeMetricName := strings.Join(metricName, ".")
+		return metricFromMap(&jsonMap, compositeMetricName, metricType)
```
Comment (gnufied) [inline src/fullerite/collector/dropwizard_collector.go]: We shouldn't really be passing pointers to maps, since variable contains just the reference anyways and passing pointers does not get anything. I followed old style, but I should probably change it.

File: src/fullerite/collector/nerve_uwsgi.go (hunk 4 above)
Comment (gnufied) [inline src/fullerite/collector/nerve_uwsgi.go]: Yeah, I will be adding them.

File: src/fullerite/collector/dropwizard_collector.go [hunk 6]
```diff
@@ -0,0 +1,162 @@
+package collector
+
+import (
+	"encoding/json"
+	"fullerite/metric"
+	"strings"
+)
+
```
Comment (tsheasha) [inline src/fullerite/collector/dropwizard_collector.go]: maybe we can add here a struct for the dropWizardJSONFormat as we are doing now with uwsgiJSONFormat struct, what do you think?

File: src/fullerite/collector/dropwizard_collector.go [hunk 7]
```diff
@@ -0,0 +1,162 @@
...
+func parseDropwizardMetric(raw *[]byte) ([]metric.Metric, error) {
+	var parsed map[string]interface{}
+
+	err := json.Unmarshal(*raw, &parsed)
```
Comment (tsheasha) [inline src/fullerite/collector/dropwizard_collector.go]: as per comment below, this is not needed, you have already unmarshalled the raw json before.

File: src/fullerite/collector/dropwizard_collector.go [hunk 8]
```diff
@@ -0,0 +1,162 @@
...
+
+	if err != nil {
+		return []metric.Metric{}, err
+	}
+
```
Comment (tsheasha) [inline src/fullerite/collector/dropwizard_collector.go]: this will be avoided as well if you pass the `parsed` to this function

File: src/fullerite/collector/dropwizard_collector.go [hunk 9]
```diff
@@ -0,0 +1,162 @@
...
+func parseMetricMap(
+	jsonMap map[string]interface{},
+	metricName []string) []metric.Metric {
+	results := []metric.Metric{}
+
+	for k, v := range jsonMap {
+		switch t := v.(type) {
+		case map[string]interface{}:
+			metricName = append(metricName, k)
+			tempResults := parseMetricMap(t, metricName)
```
Comment (tsheasha) [inline src/fullerite/collector/dropwizard_collector.go]: if we define the struct for the different metric types as mentioned in the above comment, we can get rid of this recursive behaviour and rather do a case by case parsing according to metric type. I think we can both agree that recursion is not easy for most people to understand, maybe not for yourself or myself, but maybe for others. let's have mercy :)

File: src/fullerite/collector/dropwizard_collector.go [hunk 10]
```diff
@@ -0,0 +1,162 @@
...
+			// pop the name, now that it is processed
+			if len(metricName)-1 >= 0 {
```
Comment (tsheasha) [inline src/fullerite/collector/dropwizard_collector.go]: isn't this equivalent to len(metricName) > 0

File: src/fullerite/collector/dropwizard_collector.go [hunk 11]
```diff
@@ -0,0 +1,162 @@
...
+				metricName = metricName[:(len(metricName) - 1)]
```
Comment (tsheasha) [inline src/fullerite/collector/dropwizard_collector.go]: what do you do with this metricName variable?, I don't see it being used anywhere else.

File: src/fullerite/collector/dropwizard_collector.go [hunk 12]
```diff
@@ -0,0 +1,162 @@
...
+			}
+			results = append(results, tempResults...)
+		default:
+			tempResults := parseFlattenedMetricMap(jsonMap, metricName)
```
Comment (tsheasha) [inline src/fullerite/collector/dropwizard_collector.go]: removing recursion will make this part unnecessary as well since the flattening will happen at the case level.

File: src/fullerite/collector/dropwizard_collector.go [hunk 13]
```diff
@@ -0,0 +1,162 @@
...
+				continue
+			}
+			metricType := "GAUGE"
+
+			if key == "count" {
+				metricType = "COUNTER"
+			}
+
+			compositeMetricName := strings.Join(metricName, ".")
+			m, ok := createMetricFromDatam(key, value, compositeMetricName, metricType)
```
Comment (tsheasha) [inline src/fullerite/collector/dropwizard_collector.go]: see what happened here?, we go initially from the first file nerve_uwsgi.go to this file to run `parseDropwizardMetric` then we go back to nerve_uwsgi.go to run createMetricFromDatam then we return back here to continue this function,
changing so many contexts along the way, I honestly do not think this is the way to go. As per comments before and after this I think we can merge both files and de-convolute functionality.

File: src/fullerite/collector/dropwizard_collector.go [hunk 14]
```diff
@@ -0,0 +1,162 @@
...
+
+	return results
+}
+
+func checkForRateUnits(jsonMap map[string]interface{}) bool {
+	if unit, ok := jsonMap["unit"]; ok && (unit == "seconds" || unit == "milliseconds") {
+		return true
+	}
+	return false
+}
```
Comment (tsheasha) [inline src/fullerite/collector/dropwizard_collector.go]: This function is a part of rate metric type collection, I hate to say it but it is 5 levels of scope deep a function that calls a function that calls another function .... 5 times! This IMHO is too much, especially that these sub functions are performing trivial operations and offer no code re-usability that justifies this convolution of sparse functions.

File: src/fullerite/collector/dropwizard_collector.go [hunk 15]
```diff
@@ -0,0 +1,162 @@
...
+
+func checkForMeterUnits(jsonMap map[string]interface{}) bool {
+	if _, ok := jsonMap["event_type"]; ok {
+		if unit, ok := jsonMap["unit"]; ok && (unit == "seconds" || unit == "milliseconds" || unit == "minutes") {
+			return true
+		}
+	}
+	return false
+}
```
Comment (tsheasha) [inline src/fullerite/collector/dropwizard_collector.go]: I would suggest we move the functionality of this file to the nerve_uwsgi.go file to be inside the same collector, since dropWizard is not a collector, it is just a format of metrics collected by a certain collector. This will mean that the nerve_uwsgi collector structure will be coherent with other collectors in fullerite where basically all the collector's code is self contained within one .go file.

This will also influence the structure of this current file where you will not have to move around many objects to functions by adjusting the scopes of variables and converging functions rather than convoluting when possible.

File: src/fullerite/collector/nerve_uwsgi.go [hunk 16]
```diff
@@ -223,6 +223,9 @@ func parseUWSGIMetrics(raw *[]byte) ([]metric.Metric, error) {
 	appendIt(convertToMetrics(&parsed.Histograms, metric.Gauge), "histogram")
 	appendIt(convertToMetrics(&parsed.Timers, metric.Gauge), "timer")
```
Comment (tsheasha) [inline src/fullerite/collector/nerve_uwsgi.go]: I think it will be great if we can treat converToMetrics as a black box where by we shoot at it our parsed json and it deals with all the nitty gritty of checking which json structure to match this to in order to convert it to metric, i.e. whether  python uwsgi or java services.

This would be easier if the above comments are addressed in a way where all the collector code is in one file.
Keep in mind, if you do that and make converToMetrics a black box as such you need to maintain a similar order in the calls to convert to metric as specified by the current soa collector from specific to general, i.e in the following order:
[
collect_meter,
drop_timer_entry,
drop_jvm_node,
collect_histogram,
collect_gauge,
collect_counter,
collect_rate,
]

File: src/fullerite/collector/nerve_uwsgi.go [hunk 17]
```diff
@@ -223,6 +223,9 @@ func parseUWSGIMetrics(raw *[]byte) ([]metric.Metric, error) {
...
 
+	if len(results) == 0 {
+		return parseDropwizardMetric(raw)
+	}
```
Comment (tsheasha) [inline src/fullerite/collector/nerve_uwsgi.go]: I think a better place for this check, is in `queryService` function, for semantic purposes. Since inside the current function `parseUWSGIMetrics` it is intuitive to think these metrics are from the python uwsgi module and *not* from java services *that do not run uwsgi*. Makes sense? 

Also why are you passing `raw` here instead of parsed?, this means you are unmarshalling the json twice in case it is a java service.

File: src/fullerite/collector/nerve_uwsgi.go [hunk 18]
```diff
@@ -241,26 +244,43 @@ func convertToMetrics(metricMap *map[string]map[string]interface{}, metricType s
 	results := []metric.Metric{}
```
Comment (tsheasha) [inline src/fullerite/collector/nerve_uwsgi.go]: as per the above comment we can treat this as a black box.

File: src/fullerite/collector/nerve_uwsgi.go [hunk 19]
```diff
@@ -241,26 +244,43 @@ func convertToMetrics(metricMap *map[string]map[string]interface{}, metricType s
...
-
-			// only add things that have a numeric base
-			switch value.(type) {
-			case float64:
-				m.Value = value.(float64)
-				results = append(results, m)
-			case int:
-				m.Value = float64(value.(int))
-				results = append(results, m)
-			}
```
Comment (tsheasha) [inline src/fullerite/collector/nerve_uwsgi.go]: To be honest, I am very confused with this change, correct me if I am wrong, but is the change is basically splitting up the functionality of this  function into sub-functions? If so, I believe this adds a layer of convolution that we can avoid. I see no gain in splitting up the functionality of this function, especially that this code is not being re-used anywhere and is very specific to the functionality of *that* function. Also all these temp variables we can do without if we converge this function back together. :)

I see no gain in doing this change, what do you think?

File: src/fullerite/collector/dropwizard_collector.go [hunk 20]
```diff
@@ -0,0 +1,162 @@
...
+		} else if metricType == "counter" {
+			return collectCounter(jsonMap, metricName, "counter")
+		} else if metricType == "meter" {
+			return collectMeter(jsonMap, metricName)
+		}
+	}
+
+	// if nothing else works try for rate
+	return collectRate(jsonMap, metricName)
+}
```
Comment (tsheasha) [inline src/fullerite/collector/dropwizard_collector.go]: shouldn't this replicate the behaviour of the old soa collector by running the collectors sequentially in order from specific to general and along the way dropping some jvm points as per:
[
collect_meter,
drop_timer_entry,
drop_jvm_node,
collect_histogram,
collect_gauge,
collect_counter,
collect_rate,
]

File: src/fullerite/collector/nerve_uwsgi.go (hunk 19 above)
Comment (gnufied) [inline src/fullerite/collector/nerve_uwsgi.go]: The function is being reused in dropwizard file though. I will add separate comment for whether to keep dropwizard file or not - both `metricFromMap` and `createMetricFromDatam` is being reused in several places..

File: src/fullerite/collector/dropwizard_collector.go [hunk 21]
```diff
@@ -0,0 +1,162 @@
...
+
+func collectGauge(jsonMap map[string]interface{}, metricName []string, metricType string) []metric.Metric {
+	compositeMetricName := strings.Join(metricName, ".")
+	return metricFromMap(&jsonMap, compositeMetricName, metricType)
```
Comment (tsheasha) [inline src/fullerite/collector/dropwizard_collector.go]: there is a missing check here for the key 'value' to be inside the jsonMap as per the old collector, in cases where it is not there, we must not create metrics for other items in the jsonMap as we are doing now.

File: src/fullerite/collector/nerve_uwsgi.go (hunk 17 above)
Comment (gnufied) [inline src/fullerite/collector/nerve_uwsgi.go]: That  is true - but this is because `parsed` struct can't deserialized nested stats emitted by Java services and hence I can't reuse `parsed` object.

File: src/fullerite/collector/dropwizard_collector.go [hunk 22]
```diff
@@ -0,0 +1,162 @@
...
+		return metricFromMap(&jsonMap, compositeMetricName, metricType)
+	}
+	return []metric.Metric{}
+}
+
+func collectCounter(jsonMap map[string]interface{}, metricName []string, metricType string) []metric.Metric {
+	if _, ok := jsonMap["count"]; ok {
+		compositeMetricName := strings.Join(metricName, ".")
+		return metricFromMap(&jsonMap, compositeMetricName, metricType)
+	}
```
Comment (tsheasha) [inline src/fullerite/collector/dropwizard_collector.go]: Here you missed the logic to create everything into gauges, except for the `count` entry to be a counter

File: src/fullerite/collector/dropwizard_collector.go [hunk 23]
```diff
@@ -0,0 +1,162 @@
...
+	compositeMetricName := strings.Join(metricName, ".")
+	return metricFromMap(&jsonMap, compositeMetricName, metricType)
+}
+
+func collectHistogram(jsonMap map[string]interface{}, metricName []string, metricType string) []metric.Metric {
+	if _, ok := jsonMap["count"]; ok {
+		compositeMetricName := strings.Join(metricName, ".")
+		return metricFromMap(&jsonMap, compositeMetricName, metricType)
+	}
+	return []metric.Metric{}
```
Comment (tsheasha) [inline src/fullerite/collector/dropwizard_collector.go]: Here you missed the logic to create everything into gauges, except for the `count` entry to be a counter

File: src/fullerite/collector/dropwizard_collector.go [hunk 24]
```diff
@@ -0,0 +1,162 @@
...
+}
+
+func collectCounter(jsonMap map[string]interface{}, metricName []string, metricType string) []metric.Metric {
+	if _, ok := jsonMap["count"]; ok {
+		compositeMetricName := strings.Join(metricName, ".")
+		return metricFromMap(&jsonMap, compositeMetricName, metricType)
+	}
+	return []metric.Metric{}
+}
```
Comment (tsheasha) [inline src/fullerite/collector/dropwizard_collector.go]: I couldn't find in the code whether we enforce this to be rolled up as a counter or not, but we should

File: src/fullerite/collector/nerve_uwsgi.go (hunk 17 above)
Comment (tsheasha) [inline src/fullerite/collector/nerve_uwsgi.go]: two points here:
1) parsed struct *can* deserialise these stats you just need to tell the json what to unmarshall on, see previous comment about creating a struct for the dropwizardJSONFormat.
2) You call json.Unmarshall again in parseDropwizardMetric with a generic map[string]interface{}, I am thinking in the spirit of making the detection of which metrics they are as close to the top of the funnel as possible, we can unmarshall first according to nerveUwsgiJSONFormat and if that fails we unmarshall accorsing to the dropwizardJSONFormat and that will tell us really early on in the game which service type we are doing metrics for, python uwsgi or java, and we can diverge code paths from there through the convertToMetrics "black box" approach i mentioned earlier ...  ooh I love this idea :)

File: src/fullerite/collector/dropwizard_collector.go (hunk 21 above)
Comment (gnufied) [inline src/fullerite/collector/dropwizard_collector.go]: We aren't creating a metric if `value` is missing inside `jsonMap`. If that were the case, we would be emitting multiple metrices for something like:

```
{
  type: "gauge"
  value: 100
}
```

It is just that, check happens in different way now. In new code, we only create metric for numeric fields, rather than explicitly checking for `value`.  This has same intended effect. All gauges will have numeric value with `value` key, if they do not and key is named something else but has numeric value - a metric will be created with whatever was the numeric key name. 

In practice, this is probably better approach than looking for `value` field.

File: src/fullerite/collector/nerve_uwsgi.go (hunk 19 above)
Comment (tsheasha) [inline src/fullerite/collector/nerve_uwsgi.go]: This was written with the intention of converging the files into one.

File: src/fullerite/collector/dropwizard_collector.go (hunk 20 above)
Comment (gnufied) [inline src/fullerite/collector/dropwizard_collector.go]: Not necessarily - we don't need to drop anything here because the code ignores anything that it can't parse. For example, old soa collector dropped JVM vm stat because both key and value were non-numeric and can't be reported. 

New code checks for numerical values and emits metrices only for them and hence will drop JVM vm map anyways.

File: src/fullerite/collector/dropwizard_collector.go (hunk 21 above)
Comment (tsheasha) [inline src/fullerite/collector/dropwizard_collector.go]: I beg to differ that this is better approach then looking for `value` field, which is the only field we are concerned with reporting. I think *in practice* we would rather explicitly mention the metrics we collect than have an implicit type check for the value determine whether or not we report the metric.

File: src/fullerite/collector/dropwizard_collector.go (hunk 20 above)
Comment (tsheasha) [inline src/fullerite/collector/dropwizard_collector.go]: well, 2 points here:-
1) That still does not answer why we do not run the collectors in the order specified by the old soa collector
2) jvm_node info has a version field that *can* be an int *or an integer casted from a string* and in which case will be reported wrongly as a metric.

File: src/fullerite/collector/dropwizard_collector.go (hunk 6 above)
Comment (gnufied) [inline src/fullerite/collector/dropwizard_collector.go]: I did think about it and I am more than happy to define a struct specially for dropwizard format.But, looking at unit tests and sample data, these stats are kinda irregular and do not conform to an schema like UWSGI format. 

We basically have unknown amount of nesting in json maps and any schema we define is going to be useless because of that. 

If you can think of something, I will be happy to incorporate.

File: src/fullerite/collector/dropwizard_collector.go (hunk 23 above)
Comment (gnufied) [inline src/fullerite/collector/dropwizard_collector.go]: Good catch. thanks.

File: src/fullerite/collector/dropwizard_collector.go (hunk 6 above)
Comment (tsheasha) [inline src/fullerite/collector/dropwizard_collector.go]: I have a few thoughts on this, I can hack up a quick mockup and will suggest it here, but feel free to go ahead with trying ti unravel this mystery as well, the more brains on it the better!

File: src/fullerite/collector/dropwizard_collector.go (hunk 14 above)
Comment (gnufied) [inline src/fullerite/collector/dropwizard_collector.go]: I personally hate nesting if-else conditions too much and if I had them inline inside `collectRate` method, the nesting will go crazy.

I am using a pattern called, [decomposeConditional http://refactoring.com/catalog/decomposeConditional.html] to remove complicated conditionals like these.

So this function exists to avoid a. nesting of if/else clauses b. of naming check that we are performing on metric.

File: src/fullerite/collector/dropwizard_collector.go (hunk 20 above)
Comment (gnufied) [inline src/fullerite/collector/dropwizard_collector.go]: 1. I don't know why new code should be exact replica of old code and as long as it passed same unit tests - it should be okay?

2.I think you meant - https://github.com/Yelp/fullerite/pull/81/files#diff-7b8baccceaf260e5b5be2b0c2fbf035aR827 ? I am not too much worried about it honestly. The reason, this metrices were dropped was because they were not number in old collector, if tomorrow Java version is simply a number, say `23` - who is to say, it shouldn't be reported as other JVM stats are being reported?

Comment (tsheasha) [inline src/fullerite/collector/dropwizard_collector.go]: 1) unfortunately it's not that simple, the reason it need to be an exact replica of the old code is due to how user's expect this code to behave in production. So running it against a samlle file is definitely not enough, this same reasoning is the reason why we are doing all this work actually, because we *tested it and ran it and it worked*
2) This brings me nicely to my second point,  I know "who is to say whether the java version should be reported as other JVM stats or not", our users! We can ask them about this and any ither matter we are unsure about or make assumotions of, they will surely guide us towards a better solution. <3 our users

File: src/fullerite/collector/nerve_uwsgi.go (hunk 19 above)
Comment (gnufied) [inline src/fullerite/collector/nerve_uwsgi.go]: I am leaning towards merging the files or extracting common  functions in a library as @zachm suggested. But in either case - I like these functions being separate f I am going to reuse them. if I am not going to reuse them, I will merge them into one as they were before.

File: src/fullerite/collector/dropwizard_collector.go (hunk 14 above)
Comment (tsheasha) [inline src/fullerite/collector/dropwizard_collector.go]: You can simplify the conditional in a way that wouldn't require a decomposition of it, such as checking the membership of unit in a set of valid  units, this avoids a check for unit == x || unit==y  and instead makes it something like valid[value]

File: src/fullerite/collector/nerve_uwsgi.go (hunk 19 above)
Comment (tsheasha) [inline src/fullerite/collector/nerve_uwsgi.go]: I guess all three of us are on the same page then    +1 :)

File: src/fullerite/collector/dropwizard_collector.go (hunk 20 above)
Comment (gnufied) [inline src/fullerite/collector/dropwizard_collector.go]: What old code did was, unroll the map until you arrive on flattened map and then combine all keys encountered so far to generate metrics for flattened map. This code - does exact same thing. Except minor differences:

a. I am not explicitly dropping `jvm.vm` metrices, because I don't need to! It would be trivial to drop them like how old collector did. 
b. I am not dropping `timer`  entry because that entry does not generate any metrices and hence will be ignored.
c. The gauge value you pointed. I can perhaps add a check if that makes you happy. :-)

As for order of calling different collectors, if you read old code - you know  only thing important is, old code drops jvm.vm metrics and timer entry before doing rate or other computations - other than that order is not important (they are mutually exclusive, there is no chance two metrices can be processed by same collector function).


To sum up - new code is logically same as old one. The order of calling collectors is only important for dropping jvm.vm stats and timer entry.

File: src/fullerite/collector/dropwizard_collector.go (hunk 21 above)
Comment (gnufied) [inline src/fullerite/collector/dropwizard_collector.go]: Cool, I will add that. :-)

File: src/fullerite/collector/dropwizard_collector.go (hunk 10 above)
Comment (gnufied) [inline src/fullerite/collector/dropwizard_collector.go]: Ha! agreed.

File: src/fullerite/collector/dropwizard_collector.go (hunk 11 above)
Comment (gnufied) [inline src/fullerite/collector/dropwizard_collector.go]: So basically this code (and old soa colector code), unrolled deeply nested maps/dict until - it finds a flattened map, at that point we are ready to emit metrices. 

However metric name is generated by combing keys encountered so far. For example if map was like this:

```
{
   "org.apache.foo": {
        "servicex": {
            duration: {
               p95: 4
 ```

The metric name would be `org.apache.foo.servicex.duration.p95` . Old collector did not dimenistionalize rollups but new code does, so reported metric name becomes `org.apache..duration` basically.

`metricName` is being used.

File: src/fullerite/collector/dropwizard_collector.go (hunk 13 above)
Comment (gnufied) [inline src/fullerite/collector/dropwizard_collector.go]: Yeah, I am thinking about either comibing the files or extracting out common functions. Thanks.

File: src/fullerite/collector/dropwizard_collector.go (hunk 20 above)
Comment (tsheasha) [inline src/fullerite/collector/dropwizard_collector.go]: to reply to your points

a. explicitly dropping a metric is much more clear *for ourselves and future maintainers* than to implicitly drop it due to a type check of the metric value, metrics are usually drop by the metric_name **the key**

b. you are right in that it will not generate metrics, but the code not being aware of it's existence at all is what seems odd to me, don't consider this blocking though.

c. only add it or not add it if you think it makes sense and will bring value to the module, regardless of my feeling towards it.

As for the order you are right I just noticed that you call the collectRate at the end, and this is the only one where order matters, so ignore all my previous comments on the order of collection.

File: src/fullerite/collector/dropwizard_collector.go (hunk 11 above)
Comment (tsheasha) [inline src/fullerite/collector/dropwizard_collector.go]: so the value is updated in each recursive cycle?

File: src/fullerite/collector/nerve_uwsgi.go (hunk 19 above)
Comment (gnufied) [inline src/fullerite/collector/nerve_uwsgi.go]: I have merged back dropwizard file into nerveswgi collector file.

File: src/fullerite/collector/dropwizard_collector.go (hunk 10 above)
Comment (gnufied) [inline src/fullerite/collector/dropwizard_collector.go]: Not needed anymore. Removed.

File: src/fullerite/collector/dropwizard_collector.go (hunk 12 above)
Comment (gnufied) [inline src/fullerite/collector/dropwizard_collector.go]: Recursion is gone in the version I pushed just now.

File: src/fullerite/collector/dropwizard_collector.go (hunk 14 above)
Comment (gnufied) [inline src/fullerite/collector/dropwizard_collector.go]: I have moved the condtion back to function.

**Comment by gnufied** [comment]
I have removed recursion from parsing logic and now it is kinda BFS search which uses a queue and probably it is easier to read/understand now.

I have also merged this file back into nerve_uwsgi_collector.go file, so no confusion about calls going back forth.

On negative side, now `nerve_uswgi_xx` file is big. 

I agree with @tsheasha that, it would be nice if we can decide what format data is coming in at the earliest. Right now, we can inspect UWSGI structs and if they failed to parse the JSON data, it means - it must be in Dropwizard format. But it is hacky solution at best.

Sounds like, we should implement custom `mime-type` for reporting metrices of HTTP services. Custom `mime-types` have become pretty popular for implementing RESTful web serivces, I don't see why we shouldn't use them (after all metrics reported on a HTTP endpoint is kind of an API call) for metrics reported.

File: src/fullerite/collector/nerve_uwsgi.go [hunk 25]
```diff
@@ -16,6 +16,13 @@ import (
 	l "github.com/Sirupsen/logrus"
 )
 
+const (
+	// MetricTypeCounter String for counter metric type
+	MetricTypeCounter = "COUNTER"
+	// MetricTypeGauge String for Gauge metric type
+	MetricTypeGauge = "GAUGE"
+)
+
```
Comment (tsheasha) [inline src/fullerite/collector/nerve_uwsgi.go]: What is the benefit of this, compared to having it in the code as the raw string?
In case there is a benefit that I am not aware of, it is **almost always** better inside a constant definition to define the `go` type of that constant  instead of leaving it as an un-typed constant. for more insight into what constant does under the hood ... "https://blog.golang.org/constants#TOC_4."

Comment (gnufied) [inline src/fullerite/collector/nerve_uwsgi.go]: I have gone back and fourth between using the string and the constant. Mainly I am using constant because, editors/IDEs work much better on variable names and strings are so easy to mistype, when more than once. +1 about using typed constants, i will change it.

File: src/fullerite/collector/dropwizard_collector.go (hunk 24 above)
Comment (gnufied) [inline src/fullerite/collector/dropwizard_collector.go]: It is rolled up as counter. But I am confused though, why you think it isn't?

File: src/fullerite/collector/sample.json (hunk 1 above)
Comment (gnufied) [inline src/fullerite/collector/sample.json]: I have moved this file to top level fixtures directory to make its purpose clearer. I think, having a place to store all kind of fixture data while we build up this code isn't bad idea.

**Comment by gnufied** [comment]
I think, I am all done here with various review comments.  If I have missed something let me know.

File: src/fullerite/collector/dropwizard_collector.go (hunk 24 above)
Comment (tsheasha) [inline src/fullerite/collector/dropwizard_collector.go]: :+1:  cool!

**Comment by zachm** [comment]
Signed off, @tsheasha . It's a fine point that I don't think will bite us.
